context.go

```diff
@@ -75,52 +77,69 @@ func (d RunAddressData) isEmpty() bool {
 // matches and actions can still be returned, for instance in the case of a timeout error. Errors can be tested against
 // the RunError type.
 // Struct fields having the tag `ddwaf:"ignore"` will not be encoded and sent to the WAF
-func (context *Context) Run(addressData RunAddressData, timeout time.Duration) (res Result, err error) {
+// if the output of TotalTime() exceeds the value of TimeBudget, the function will immediately return with an error
+func (context *Context) Run(addressData RunAddressData, _ time.Duration) (res Result, err error) {
 	if addressData.isEmpty() {
-		return
+		return res, nil
 	}
 
-	now := time.Now()
+	// Everything here is thread safe and ran after context.mutex.Unlock() so we lock the context the less possible
+	// so the code here needs to be thread safe
 	defer func() {
-		dt := time.Since(now)
-		context.totalOverallRuntimeNs.Add(uint64(dt.Nanoseconds()))
+		if stdErrors.Is(err, errors.ErrTimeout) {
+			context.timeBudgetExpirationCount.Inc()
+		}
 	}()
 
-	// At this point, the only error we can get is an error in case the top level object is a nil map, but this
-	// behaviour is expected since either persistent or ephemeral addresses are allowed to be null one at a time.
-	// In this case, EncodeAddresses will return nil contrary to Encode which will return an nil wafObject,
-	// which is what we need to send to ddwaf_run to signal that the address data is empty.
-	var persistentData *bindings.WafObject = nil
-	var ephemeralData *bindings.WafObject = nil
-	persistentEncoder := newLimitedEncoder()
-	ephemeralEncoder := newLimitedEncoder()
-	if addressData.Persistent != nil {
-		persistentData, _ = persistentEncoder.EncodeAddresses(addressData.Persistent)
+	// If the context has already timed out, we don't need to run the WAF again
+	if context.timings.Expired() {
+		return res, errors.ErrTimeout
 	}
 
-	if addressData.Ephemeral != nil {
-		ephemeralData, _ = ephemeralEncoder.EncodeAddresses(addressData.Ephemeral)
+	runTimings := context.timings.newRun()
+	startTime := time.Now()
+	defer func() {
+		runTimings.RunNs.Add(time.Since(startTime).Nanoseconds())
+		context.timings.absorb(&runTimings)
+	}()
 
+	persistentData, persistentEncoder, err := context.encodeOneAddressType(addressData.Persistent, &runTimings)
+	if err != nil {
+		return res, err
 	}
-	// The WAF releases ephemeral address data at the end of each run call, so we need not keep the Go values live beyond
+
+	// The WAF releases ephemeral address data at the max of each run call, so we need not keep the Go values live beyond
 	// that in the same way we need for persistent data. We hence use a separate encoder.
+	ephemeralData, ephemeralEncoder, err := context.encodeOneAddressType(addressData.Ephemeral, &runTimings)
+	if err != nil {
+		return res, err
+	}
```

Comment:
If we time out during encode, but we successfully encoded _something_, are we returning an error, or are we proceeding with partial input?

I feel like proceeding with partial input might be preferable (we spent time encoding those things, this is pure sunk cost unless we actually use that).

If we process partial input, it might be worth wondering if we should encode persistent or ephemeral data first?